[IPv6][SAM] Enable multicast hash table for multiple MAC addresses on SAM network interface

ActoryOu opened the pull request:
<!--- Title -->
[IPv6][SAM] Enable multicast hash table for multiple MAC addresses on SAM network interface

Description
-----------
<!--- Describe your changes in detail. -->
The change is made by @htibosch.

When IPv6/Multiple endpoints enabled, the space to store MAC addresses is not enough.
To support it better, use multicast hash table in hardware to reduce loading on space.

Test Steps
-----------
<!-- Describe the steps to reproduce. -->
Below scenarios are tested (by @htibosch) on both IPv4/IPv6:
- DNS
- UDP
- TCP
- Ping
- IPERF3

Checklist:
----------
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have tested my changes. No regression in existing tests.
- [x] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------
<!-- If any, please provide issue ID. -->


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

kar-rahul-aws commented on source/portable/NetworkInterface/DriverSAM/NetworkInterface.c:
```diff
@@ -506,6 +510,8 @@ NetworkInterface_t * pxSAM_FillInterfaceDescriptor( BaseType_t xEMACIndex,
     pxInterface->pfOutput = prvSAM_NetworkInterfaceOutput;
```
When setting the pcName , it can be set as 
```
snprintf( pcName, sizeof( pcName ), "GMAC%u", ( unsigned ) xEMACIndex );
pxInterface->pcName = pcName;
```
This method is being followed for STM32Hxx port, and might be more useful for debugging purpose.

ActoryOu replied:
Updated

kar-rahul-aws commented on source/portable/NetworkInterface/DriverSAM/NetworkInterface.c:
```diff
@@ -680,9 +686,18 @@ static BaseType_t prvGMACInit( NetworkInterface_t * pxInterface )
     #if ( ipconfigUSE_LLMNR == 1 )
         {
-            gmac_set_address( GMAC, xEntry++, xLLMNR_MacAdress.ucBytes );
+            prvAddMACAddress( xLLMNR_MacAdress.ucBytes );
```
Extra check is needed that ipconfigUSE_IPv6  == 0.

ActoryOu replied:
We shouldn't check `ipconfigUSE_IPv6 == 0` here. This mac address is used for IPv4 LLMNR, which is independent with IPv6.

tony-josi-aws commented on source/portable/NetworkInterface/DriverSAM/NetworkInterface.c:
```diff
@@ -691,7 +710,7 @@ static BaseType_t prvGMACInit( NetworkInterface_t * pxInterface )
             NetworkEndPoint_t * pxEndPoint;
```
Maybe its more cleaner (or more readable) to have different names for variables (`pxEndPoint`) though they are in different scope.

htibosch commented:
Thank you Actory for creating this PR.
What I miss is a few items in the description of the PR:

- The name of the NetworkInterface will change from "GMAC" to "GMAC1", depending on the assigned index.
- The function `pxSAM_FillInterfaceDescriptor()` now calls `FreeRTOS_AddNetworkInterface()`, as it should
- Adapted `vGMACGenerateChecksum()` for use with IPv6 addresses
- An event passed to the IP-task does not have to be declared static as in `static IPStackEvent_t xRxEvent`
- When `FreeRTOS_MatchingEndpoint()` does not find a match, drop the received packet

Some remarks:

The function `prvAddMACAddress()` is too generic. Maybe we should call it `prvAddMulticastMACAddress()`
At the bottom there is the function `gmac_enable_irq()`. I used it for testing: it makes stepping easier. You can remove the declaration.

ActoryOu commented:
> gmac_enable_irq



> Thank you Actory for creating this PR. What I miss is a few items in the description of the PR:
> 
> * The name of the NetworkInterface will change from "GMAC" to "GMAC1", depending on the assigned index.
> * The function `pxSAM_FillInterfaceDescriptor()` now calls `FreeRTOS_AddNetworkInterface()`, as it should
> * Adapted `vGMACGenerateChecksum()` for use with IPv6 addresses
> * An event passed to the IP-task does not have to be declared static as in `static IPStackEvent_t xRxEvent`
> * When `FreeRTOS_MatchingEndpoint()` does not find a match, drop the received packet
> 
> Some remarks:
> 
> The function name `prvAddMACAddress()` is too generic. Maybe we should call it `prvAddMulticastMACAddress()` At the bottom there is the function `gmac_enable_irq()`. I used it for testing: it makes stepping easier. You can remove the declaration.

Thanks for additions.
I've updated `prvAddMACAddress` to `prvAddMulticastMACAddress` and removed `gmac_enable_irq`.

Please help to take a look.
Thanks.

tony-josi-aws commented on source/portable/NetworkInterface/DriverSAM/NetworkInterface.c:
```diff
@@ -706,7 +725,7 @@ static BaseType_t prvGMACInit( NetworkInterface_t * pxInterface )
                     ucMACAddress[ 3 ] = pxEndPoint->ipv6_settings.xIPAddress.ucBytes[ 13 ];
```
When adding IPv6 multi-cast MAC addresses of the IPv6 endpoints to the hash table, shouldn't it be of the form - [Address Mapping -- Multicast](https://datatracker.ietf.org/doc/html/rfc2464#section-7)

shubnil replied:
This looks like a valid issue and the field ucMACAddress[ 2 ] needs to be set as well.

htibosch replied:
Change will be reverted, see [Solicited-node multicast address](https://en.wikipedia.org/wiki/Solicited-node_multicast_address). It is used to give the DUT a multicast address which is used during the SLAAC for DAD (Duplicate Address Detection).
